trigger workspace/didChangeConfiguration on compilation database modification

BenjaminNavarro opened the pull request:
I was annoyed to always have to call `Restart language server`or `Restart language server in lazy mode` manually when my compilation database was modified so here is my attempt to fix this.

I simply use `workspace.createFileSystemWatcher` to monitor changes to the `compilation_database.json` file and when a change occur, I send the `workspace/didChangeConfiguration` notification to the server.

It seems to be working for me but since it's the first time I work on a VSCode extension I might have missed a couple things.

Yanpas committed:
dataflow

Yanpas committed:
show explorer

Yanpas committed:
merge with master

MaskRay committed:
0.1.22

MaskRay committed:
Rename cacheDirectory to cache.directory

MaskRay committed:
0.1.23

Riatre committed:
Add option ccls.codeLens.enabled

This allows to disable CodeLens extension-wise. Some users may want to see CodeLens provided by other extensions, thus they do not want to disable CodeLens globally.

Riatre committed:
Add option ccls.trace.websocketEndpointUrl to log LSP traffic to WebSocket endpoint (#51)

This can be used with lsp-inspector-webview
(https://marketplace.visualstudio.com/items?itemName=octref.lsp-inspector-webview)
for easier debugging.

MaskRay committed:
Add .github/ISSUE_TEMPLATE

Riatre committed:
Cleanup config mapping logic in getClientConfig

Populate ClientConfig with all keys from VS Code WorkspaceConfiguration
(except those in blacklist) instead of a preset list.

With this change, it is possible to send initializationOptions to ccls
even if there is no corresponding configuration declaration in
package.json. So new initializationOption exposed in ccls no longer
necessarily reuqires an update of vscode-ccls.

Declarations in package.json are still useful as they still provide
documentations and default values.

Also no longer add workspaceSymbol.sort = false by default as this works
poorly when there are too many symbols.

Riatre committed:
Add documentation of index.maxInitializerLines; set default value to 15

MaskRay committed:
Use the default argument of child_process.spawn

  const defaults = {
    cwd: undefined,
    env: process.env
  };

Deokkoo committed:
Add option ccls.callHierarchy.qualified to show detailed names in the call hierarchy

Defaults to true.

theorlangur committed:
fixed/improved member hierarchy info

now it shows names of fields as well as their types
+ byte offset of particular field in the tooltip

MaskRay committed:
npm update

MaskRay committed:
Redesign semantic highlight configurations (#75)

Adopt the emacs-ccls approach.

ccls.highlight.{function,type,type}.face define faces of basic symbol
kinds from which other symbol kinds
({member,staticMember}{Function,Variable}, etc) can inherit.

For example, if

    ccls.highlight.function.face: ["enabled"]

is specified, memberFunction will also be enabled because of the default
configurations:

    // Inherit from both ccls.highlight.{function,member}.face
    ccls.highlight.memberFunction.face: ["function", "member"]
    ccls.highlight.member.face: ["fontStyle: italic"]

For a symbol kind, its face property takes effect if "enabled" is
specified.

MaskRay committed:
Bump to 0.1.28 and refactor packaging

BenjaminNavarro committed:
trigger workspace/didChangeConfiguration on compilation database modification

BenjaminNavarro committed:
coding sytle

BenjaminNavarro commented:
Done. Could you please explain me the rationale between using single or double quotes? Both seem to be used in the code so I'd like to know when each one has to be used in the case I have other modification to make

BenjaminNavarro commented:
When can this be merged?

MaskRay commented:
If the initialization option `compilationDatabaseDirectory` is not empty, it specifies the directory that contains `compile_commands.json`. Can you update the patch to recognize  `compilationDatabaseDirectory`?

```cpp
// ccls/src/project.cc
    cdbDir = root;
    if (g_config->compilationDatabaseDirectory.size()) {
      if (sys::path::is_absolute(g_config->compilationDatabaseDirectory))
        cdbDir = g_config->compilationDatabaseDirectory;
      else
        sys::path::append(cdbDir, g_config->compilationDatabaseDirectory);
    }
    sys::path::append(path, cdbDir, "compile_commands.json");
```

BenjaminNavarro committed:
Use compilationDatabaseDirectory if specified

BenjaminNavarro commented:
That should do it

MaskRay commented on src/serverContext.ts:
```diff
@@ -292,6 +292,18 @@ export class ServerContext implements Disposable {
+
+    if (config.get('index.reloadDatabaseOnChange', true)) {
+      let db_dir = config.get('misc.compilationDatabaseDirectory');
+      if (!db_dir || db_dir === '') {
```
Delete the surrounding `{}`.

BenjaminNavarro replied:
Ok of you prefer it that way

MaskRay commented on src/serverContext.ts:
```diff
@@ -292,6 +292,18 @@ export class ServerContext implements Disposable {
+      if (!db_dir || db_dir === '') {
+        db_dir = this.cwd;
+      }
+      const db_watcher = workspace.createFileSystemWatcher(db_dir + "/compile_commands.json", false, false, false);
```
`"` -> `'`

I think single quotes are generally preferred, but for some historical reasons double quotes are still used in these files...

BenjaminNavarro replied:
Wouldn't be better to change all quotes in the project once and for all and stick to that?

BenjaminNavarro committed:
coding style

BenjaminNavarro commented:
Is it ok now?

BenjaminNavarro committed:
feat: resolve possible symlinks before db modification watching

Watching a symlink instead of the target file won't actually trigger a
db reloading if the symlinked file changes

BenjaminNavarro committed:
feat: integrate dataflow PR from official project

Trass3r commented:
Looks like this somehow picked up commits from master?

GitMensch commented:
Should this be rebased to be merged?
